contracts/BucketAuction.sol

```diff
@@ -0,0 +1,282 @@
+        string memory tokenURISuffix,
+        uint256 maxMintableSupply,
+        uint256 globalWalletLimit,
+        address cosigner,
+        uint256 minimumContributionInWei
+    )
+        ERC721M(
+            collectionName,
+            collectionSymbol,
+            tokenURISuffix,
+            maxMintableSupply,
+            globalWalletLimit,
+            cosigner
+        )
+    {
+        _minimumContributionInWei = minimumContributionInWei;
+    }
+
+    modifier isAuctionActive() {
+        if (!_auctionActive) revert BucketAuctionNotActive();
+        _;
+    }
+
+    modifier isAuctionInactive() {
+        if (_auctionActive) revert BucketAuctionActive();
+        _;
+    }
+
+    function getMinimumContributionInWei() external view returns (uint256) {
+        return _minimumContributionInWei;
+    }
+
+    function getPrice() external view returns (uint256) {
+        return _price;
+    }
+
+    function getAuctionActive() external view returns (bool) {
+        return _auctionActive;
+    }
+
+    function getUserData(address user) external view returns (User memory) {
+        return _userData[user];
+    }
+
+    /**
+     * @notice begin the auction.
+     * @dev cannot be reactivated after price has been set.
+     * @param b set 'true' to start auction, set 'false' to stop auction.
+     */
+    function setAuctionActive(bool b) external onlyOwner cannotMint {
+        if (_price != 0) revert PriceHasBeenSet();
+        _auctionActive = b;
+    }
+
+    /**
+     * @notice place a bid in ETH or add to your existing bid. Calling this
+     *   multiple times will increase your bid amount. All bids placed are final
+     *   and cannot be reversed.
+     */
+    function bid() external payable isAuctionActive nonReentrant cannotMint {
+        User storage bidder = _userData[msg.sender]; // get user's current bid total
+        uint256 contribution_ = bidder.contribution; // bidder.contribution is uint216
+        unchecked {
+            // does not overflow
+            contribution_ += msg.value;
+        }
+        if (contribution_ < _minimumContributionInWei)
+            revert LowerThanMinBidAmount();
+        bidder.contribution = uint216(contribution_);
+        emit Bid(msg.sender, msg.value, contribution_, address(this).balance);
+    }
+
+    /**
+     * @notice set the minimum contribution required to place a bid
+     * @dev set this price in wei, not eth!
+     * @param minimumContributionInWei new price, set in wei
+     */
+    function setMinimumContribution(uint256 minimumContributionInWei)
+        external
+        onlyOwner
+    {
+        _minimumContributionInWei = minimumContributionInWei;
+        emit SetMinimumContribution(minimumContributionInWei);
+    }
+
+    /**
+     * @notice set the clearing price after all bids have been placed.
+     * @dev set this price in wei, not eth!
+     * @param priceInWei new price, set in wei
+     */
+    function setPrice(uint256 priceInWei)
```

Comment:
Should we only allow setting price once? 

Just read through the contract. I think the one vulnerability is allowing owner changing price. I am not sure if this is ok but just want to raise this.

e.g. owner can reset price so he can send more tokens to specific bidders.

Reply:
All refund / token calculation are based on current price. Some users may lose fund if their `refundClaimed` is set with different price.

Reply:
maybe we make it impossible to change price once there's refund happened, good catch!

Reply:
Resolved.